### ISSUE TITLE
[Branch Error] "Branch User Session has not been initialized" and "Missing session items!"

### ISSUE THREAD

**Issue by jschwendt** [BranchMetrics/iOS-Deferred-Deep-Linking-SDK#227]
I'm having issues using SDK version 0.11.9.  It appears that the init block now has to finish before any other branch calls can be made.  We've been using the SDK for several months now without issue, but this latest SDK seems to have different expectations.

Currently we call initSessionWithLaunchOptions: as one of the first things in application: didFinishLaunchingWithOptions:, but we do so in a non-blocking manor where the UI is built and other services are initialized while the Branch init is finishing.  We may call setIdentity: before initSessionWithLaunchOptions: has finished for example.  We typically call it again later to confirm.

With this latest SDK, we're now getting errors on the console like: "Branch User Session has not been initialized" and "Missing session items!".  It's a fairly large restructuring of our code to adjust to wrapping the start of the entire app within the branch call in a blocking manor.

Can you explain why this change was made if it is indeed on purpose, or is this perhaps an unintended bug recently introduced?  We ran into similar issues several months ago where we were calling userCompletedAction: before init had completed and found similar issues, thus had to comment out that piece.  It was our understanding that you were going to add better handling to the SDK to ensure the init was completed instead of throwing these errors.  We also discussed providing an isInitialized method which could be called to check if Branch is ready for additional commands or not.  This would likely be the easiest fix of all, as it would be easy for us to check this flag before making additional calls.  We've run into similar issues with the Javascript SDK, which really needs this call as well.

Please let me know your thoughts on how to best address this ASAP.  We're looking to take advantage of the new Universal links and are thus trying to use the latest SDK, but we're blocked on this issue.

**Comment by derrickstaten** [comment]
@jschwendt you can safely revert to 0.11.8. This was not intended. We'll investigate and then cut a new release if necessary. Thanks for letting us know.

**Comment by derrickstaten** [comment]
@jschwendt I'm trying various scenarios and I'm unable to trigger this. Could you turn on setDebug and send us the logs? In the meantime, we'll revert the change that we believe is causing this.

**Comment by derrickstaten** [comment]
@jschwendt we've removed the recent change made to this code. Please try `0.11.10`.

Also, can you send us (via support@branch.io if you'd prefer to to not have this public)
1. the last version of the SDK you're using
2. the location of the calls to setIdentity -- inside which methods on the app delegate and view controller(s)
3. your implementations of `application:didFinishLaunchingWithOptions:` and `application:openURL:sourceApplication:annotation:`
4. the logs for the first 10 seconds of your application running, scrubbed of any private keys or otherwise private information.

We just removed a fix for a race condition that works for some partners. We want to test that fix with you and reach out to a few other partners to have them test. Sorry for the headache -- we never want to make breaking changes, certainly not unannounced.

**Comment by jschwendt** [comment]
Thanks for the quick reply @derrickstaten!  I tried 0.11.10 and see no change in behavior unfortunately.  I'm calling initSessionWithLaunchOptions: and then immediately thereafter calling setIdentity before the init callback has completed.  Thus I'm guessing we're running into issues.  Because the call to setIdentity calls getInstance before it's inited, I see [self initUserSessionAndCallCallback:NO]; being called.  I think you need to add a self.isInitializing flag, which will handle these race conditions.  Currently you're only tracking if initialization is complete.  By blocking or queuing secondary requests while initializing, we can ensure things complete in the correct order.  Let me know if this helps or not.  If not, I'll send you an email to support with more info.  I'm not at liberty to share my code, but I can do a screen sharing session if necessary.

**Comment by derrickstaten** [comment]
@jschwendt that's the only recent change to initialization. Can you tell us what version you were on before that worked? 

If we could do a screen sharing session and you can show me the code, that'd be great. I'm sending you an email now.

**Comment by derrickstaten** [comment]
This is fixed -- a bug in 0.11.8 introduced this briefly.
